cdap-app-fabric/src/main/java/co/cask/cdap/internal/profile/ProfileService.java

```diff
@@ -155,23 +169,73 @@ public void createIfNotExists(ProfileId profileId, Profile profile) {
+   * 2. There must entities assigned to this profile, now the assignment comes from Preference Store and Schedule store
+   * 3. There must be no active program runs using this profile
    *
    * @param profileId the id of the profile to delete
    * @throws NotFoundException if the profile is not found
    * @throws ProfileConflictException if the profile is enabled
    */
   public void deleteProfile(ProfileId profileId) throws NotFoundException, ProfileConflictException {
     Transactionals.execute(transactional, context -> {
-      getProfileDataset(context).deleteProfile(profileId);
+      ProfileDataset profileDataset = getProfileDataset(context);
+      Profile profile = profileDataset.getProfile(profileId);
+      if (profile == null) {
+        throw new NotFoundException(profileId);
+      }
+
+      // The profile status must be DISABLED
+      if (profile.getStatus() == ProfileStatus.ENABLED) {
+        throw new ProfileConflictException(
+          String.format("Profile %s in namespace %s is currently enabled. A profile can " +
+                          "only be deleted if it is disabled", profileId.getProfile(), profileId.getNamespace()),
+          profileId);
+      }
+
+      // There must be no assigments to this profile
+      Set<EntityId> associations = profileDataset.getProfileAssignment(profileId);
+      if (!associations.isEmpty()) {
+        throw new ProfileConflictException(
+          String.format("There are still entities %s assigned to this profile %s.Please remove the assignment from " +
+                          "the preferenes or schedule properties to delete the profile.",
+                        associations, profileId.toString()),
+          profileId);
+      }
+
+      // There must be no running programs using the profile
+      AppMetadataStore appMetadataStore = AppMetadataStore.create(cConf, context, datasetFramework);
+      NamespaceMDS nsStore = NamespaceMDS.getNamespaceMDS(context, datasetFramework);
+      Set<NamespaceId> namespaceIds;
+      if (profileId.getNamespaceId().equals(NamespaceId.SYSTEM)) {
+        namespaceIds = nsStore.list().stream().map(NamespaceMeta::getNamespaceId).collect(Collectors.toSet());
+      } else {
+        namespaceIds = Collections.singleton(profileId.getNamespaceId());
+      }
+      Map<ProgramRunId, RunRecordMeta> activeRuns = appMetadataStore.getActiveRuns(namespaceIds, runRecordMeta -> {
+        // the profile comes in system arguments with the scoped name
+        String scopedName = runRecordMeta.getSystemArgs().get(SystemArguments.PROFILE_NAME);
+        return scopedName != null && scopedName.equals(profileId.getScopedName());
+      });
+      if (!activeRuns.isEmpty()) {
+        throw new ProfileConflictException(
+          String.format("There are still programs %s running using the profile %s. " +
+                          "Please stop them to delete the profile.", activeRuns.keySet(), profileId.toString()),
```

Comment:
this can be a large set and a long list of program runids isn't very user friendly. I would change this to just show the number of runs, and the first run in the set. Something like.

The profile is in use by %d active runs, like %s. Please stop all active runs before deleting the profile.